src/tqec/computation/open_graph.py

```diff
@@ -0,0 +1,230 @@
+    ext_stabs = [s.external_stabilizer_on_graph(graph) for s in correlation_surfaces]
+    stab_to_surface: dict[str, CorrelationSurface] = {
+        stab: surface for stab, surface in zip(ext_stabs, correlation_surfaces)
+    }
+    identity = "I" * num_ports
+    # Need to collect all the possible correlation surfaces because we want
+    # to find the generators with the smallest correlation surface area
+    for stabilizer, comb in _iter_stabilizer_group(ext_stabs):
+        if stabilizer != identity and stabilizer not in stab_to_surface:
+            correlation_surface = reduce(
+                lambda a, b: a ^ b,
+                [stab_to_surface[s] for s in comb],
+            )
+            stab_to_surface[stabilizer] = correlation_surface
+
+    # Sort the stabilizers by its corresponding correlation surface's area to
+    # find the generators with the smallest area
+    stabs_ordered_by_area = sorted(
+        stab_to_surface.keys(),
+        key=lambda s: (stab_to_surface[s].area(), s),
+    )
+    # find a complete set of generators, starting from the smallest area
+    generators: list[str] = stabs_ordered_by_area[:1]
+    generators_stim: list[stim.PauliString] = [stim.PauliString(generators[0])]
+    for stabilizer in stabs_ordered_by_area[1:]:
+        pauli_string = stim.PauliString(stabilizer)
+        if not _can_be_generated_by(pauli_string, generators_stim):
+            generators.append(stabilizer)
+            generators_stim.append(pauli_string)
+        if len(generators) == num_ports:
+            break
+
+    # Two stabilizers are compatible if they can agree on the supported observable
+    # basis on the common ports. We can construct a graph that assigns a node to
+    # each stabilizer and an edge between two nodes if the stabilizers are compatible.
+    # Then we can solve the minimum clique cover problem to find the minimum
+    # number of cliques that cover all the nodes in the graph. Each clique will
+    # correspond to a port configuration that can be used for simulation.
+    # Here we solve the graph coloring problem on the complement graph, which is
+    # equivalent to the clique cover problem on the original graph.
+    nodes = list(range(len(generators)))
+    g: nx.Graph[int] = nx.Graph()
+    g.add_nodes_from(nodes)
+    for i, j in combinations(nodes, 2):
+        if not _is_compatible_paulis(generators[i], generators[j]):
+            g.add_edge(i, j)
+    # Solve with heuristic greedy coloring
+    coloring = nx.algorithms.coloring.greedy_color(g)
+    cliques: dict[int, list[str]] = {}
+    for node, color in coloring.items():
+        cliques.setdefault(color, []).append(generators[node])
+
+    def ports_basis_for_clique(
+        supported_stabilizers: list[str],
+    ) -> list[str]:
+        port_ops = ["I"] * num_ports
+        for s in supported_stabilizers:
+            for i, p in enumerate(s):
+                if p != "I":
+                    port_ops[i] = p
```

Comment:
Just to check that I understood the algorithm, the following should be valid right?

```suggestion
            for i, p in enumerate(s):
                if p != "I":
                    assert port_ops[i] in ["I", p]
                    port_ops[i] = p
```

Reply:
That's correct.